goalie/adjoint.py

```diff
@@ -622,7 +622,7 @@ def wrapped_solver(subinterval, initial_condition_map, **kwargs):
 
             # Get adjoint action on each subinterval
             with pyadjoint.stop_annotating():
-                for field, control in zip(self.fields, self._controls):
+                for field, control in zip(self.field_data, self._controls):
```

Comment:
This looks a bit fragile: you're relying on the order of `field_data` and `_controls` to be the same and it's very hard to work out from the code whether it is (and relying on the order in a dictionary is always a bit iffy)

Reply:
Good point! The controls should also be in a dictionary.

Reply:
Addressed in 3f6cc7b.